data/tudu.1
.\"  Copyright (C) 2007-2012 Ruben Pollan Bella <meskio@sindominio.net>         
.\"  Copyright (C) 2010 Enrique Matías Sánchez <cronopios@gmail.com>
  
.\"                                                                        
.\"   This file is part of TuDu.                                           
.\"                                                                        
.\"   TuDu is free software; you can redistribute it and/or modify         
.\"   it under the terms of the GNU General Public License as published by 
.\"   the Free Software Foundation; version 3 of the License.       
.\"                                                                        
.\"   TuDu is distributed in the hope that it will be useful,              
.\"   but WITHOUT ANY WARRANTY; without even the implied warranty of       
.\"   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the        
.\"   GNU General Public License for more details.                         
.\"                                                                        
.\"   You should have received a copy of the GNU General Public License    
.\"   along with this program.  If not, see <http://www.gnu.org/licenses/>.
.\"
.TH TuDu 1 "January 09, 2010" Unix "User Manuals"
.SH NAME
TuDu \- A command-line tool to manage TODO lists hierarchically.
.SH SYNOPSIS
.PP 
\fBtudu \fR[\fB-f \fItudu_file\fR] [\fB-c \fIconfig_file\fR] [\fB-v[v]\fR]
[\fB-h\fR]
.SH DESCRIPTION
.PP
\fBTuDu\fR is a commandline tool to manage hierarchical TODO lists, so that you
can organize everything you have to do in a simple and efficient way. 
It does not use any database backend, but plain XML files.

Every TODO is composed by:
.TS
tab (>);
l l.
Status>Currently only two options: Pending and Done.
Title>A short sentence summarizing the task.
Description text>A longer text providing more details. (Optional)
Priority>The urgency of the task. (Optional)
Category>A short tag such as @home, @work, @inet. (Optional)
Deadline>The latest date the task should be done. (Optional)
Scheduled date>The planned date to do the task. (Optional)
Subtasks>Every task can have any number of subtasks. (Optional)
.TE

Tasks can be ordered by any of those fields, and categories can be used to show
only some certain kinds of tasks. Tudu also warns when a deadline is
approaching, and shows the percentage of the task done so far.
.SH OPTIONS
.IP "-f \fItudu_file\fP" 10
Load the given tudu file
.IP "-c \fIconfig-file\fP" 10
Load specific config file
.IP "-v" 10
Display the TuDu version number.
.IP "-vv" 10
Display license and copyright information.
.IP "-h" 10
Display help.
.SH INTERFACE
All the key bindings described here are the defaults in tudu. They can be
changed in the config file.
.IP h
Move the cursor one level out in the tree depth.
.IP l
Move the cursor one level in in the tree depth.
.IP j
Move the cursor to the next task.
.IP k
Move the cursor to the previous task.
.IP J
Move a task one position down.
.IP K
Move a task one position up.
.IP m
Mark a task as done.
.IP dd
Delete a task.
.IP dt
Remove the deadline of the task.
.IP df
Remove the priority of the task.
.IP dS
Remove the scheduled date of the task.
.IP p
Paste the last deleted task under the current cursor line.
.IP P
Paste the last deleted task over the current cursor line.
.IP V
Paste the last deleted task as a subtask of the currently selected task.
.IP o
Add new task.
.IP O
Add a new task over the current cursor line.
.IP a
Edit the title of the task.
.IP t
Edit the deadline.
.IP f
Set the priority of the task.
.IP C
Add or modify the category of the task.
.IP e
Edit the description text of the task.
.IP x
Scroll the text window upwards.
.IP z
Scroll the text window downwards.
.IP S
Edit the scheduled date of the task.
.IP X
Scroll the schedule window upwards.
.IP Z
Scroll the schedule window downwards.
.IP c
Collapse or expand the children of the current task.
.IP M
Hide or show tasks marked as done.
.IP /
Search for a pattern in tasks titles.
.IP n
Search the next occurrence of last search.
.IP N
Search the previous occurrence of the last searched tasks.
.IP /
Search for a pattern in tasks titles.
.IP n
Search the next occurrence of last search.
.IP N
Search the previous occurrence of the last search.
.IP :
Open the command input. See the COMMANDS section.
.IP bt
Sort by title.
.IP bT
Sort by title in reverse order.
.IP bd
Sort by done.
.IP bD
Sort by done in reverse order.
.IP bl
Sort by deadline.
.IP bL
Sort by deadline in reverse order.
.IP bp
Sort by priority.
.IP bP
Sort by priority in reverse order.
.IP bc
Sort by category.
.IP bC
Sort by category in reverse order.
.IP bu
Sort by user-defined order.
.IP bU
Sort by reversed user-defined order.
.IP bd
Sort by done.
.IP bD
Sort by done in reverse order.
.IP bl
Sort by deadline.
.IP bL
Sort by deadline in reverse order.
.IP bp
Sort by priority.
.IP bP
Sort by priority in reverse order.
.IP bc
Sort by category.
.IP bC
Sort by category in reverse order.
.IP bu
Sort by user.
.IP bU
Sort by user in reverse order.
.IP s
Save all the changes.
.IP ?
Display the help.
.IP q
Quit tudu saving the last changes.
.IP Q
Quit tudu without saving anything.
.SH COMMANDS
.IP "hide category1 category2 ... categoryN"
Hide one or more categories.
.IP "show category1 category2 ... categoryN"
Display one or more categories.
.IP showall
Display all the categories.
.IP "showonly category1 category2 ... categoryN"
Hide all the categories except the ones given.
.IP "help"
Display the man page.
.SH CONFIG
The config file have several sections which start by a name between brackets,
like \fB[ general ]\fR. Other config files can be included with the option:
.IP "@include = 'path'"

.PP
The options of each section are:

.SS general
.IP "collapse = yes|\fBno\fR"
Collapse the tasks by default.
.IP "hide_done = yes|\fBno\fR"
Whether to hide tasks marked as done.
.IP "hide_percent = yes|\fBno\fR"
Whether to hide the percentage display on each task.
.IP "visual_tree = yes|\fBno\fR"
Toggle graphical tree. It shows marks, like \fB(+)\fR at the left of the tasks.
It is necessary in some shells that can not display bold characters.
.IP "bold_parent = \fByes\fR|no"
Hilight the parents with children with bold letters.
.IP "loop_move = yes|\fBno\fR"
Whether to jump to the beginning (or end) when the end (or beginning) is reached.
.IP "days_warn = number"
How many days before the task is due will the deadline sport a warning remark.
By default is 7.
.IP "us_dates = yes|\fBno\fR"
Use US date format (mm/dd/yyyy), by default it uses dd/mm/yyyy.
.IP "old_sched = \fByes\fR|no"
Whether display or not past scheduled tasks on the scheduler. In case of yes the
scheduled tasks for days already past will be displayed on warn color.
.IP "tudu_file = full-path"
Location of the tudu xml to be load if no '-f' parameter is given. The path have
to be a full path. If is not set ~/.tudu.xml will be use.
.IP "sort_order = string"
Default order in which tasks will be shown. Tasks are ordered by the 1st letter
in the string, then by the 2nd, and so on.

The letters mean:
.RS
.IP t
title
.IP T
reverse title
.IP d
done
.IP D
reverse done
.IP l
deadline
.IP L
reverse deadline
.IP c
category
.IP C
reverse category
.IP e
percentage done
.IP E
reverse percentage done
.IP u
user-defined order
.IP U
reverse user-defined order
.RE
.IP "editor = 'path_bin %s'"
.SS keys
This section defines any key binding on the program. See the example in the
default config file.
.SS theme
Window positions are described by rows; each row with the syntax:
.RS
row = [height]([width|]window[, [width|]window, ...])
.RE
The height and width can be the number of characters or a percentage followed by
\fB%\fR. The possible windows are: help, tree, info, blank, vpipe, hpipe text, 
and schedule. The windows help, vpipe and info don't need to have height, neither 
hpipe needs to have width, because is predefined as 1.
TuDu will try to guess the non given information.

Other options for the theme are:
.IP "columns = colm1,...,colmN"
Columns in tree window. The possible columns are: title, priority, category and
deadline.
.IP "category_length = numChar"
Number of characters to be displayed of the category.
.IP "color = color_text, color_background"
The default colors for the text and background. If your terminal has support to 
change colora the colors can be defined  as \fB(r, g, b)\fR, where r, g and b are
numbers from 0 to 1000, or with the words: black, red, green, yellow, blue, 
magenta, cyan, white, transparent.
.IP "window = color_text, color_background"
For each window can be defined it's own colors. The colors of vpipe and hpipe are
the same, and defined with the window name \fBpipe\fR.
.IP "selected = color_text, color_background"
Color for the items under the cursor.
.IP "warn = color_text, color_background"
Color for the warnings, like the mark of deadline soon or the color of the task
scheduled on the past.
.SH FILES
.PP
.IP "~/.tudurc"
User configuration file.
.IP "/etc/tudurc"
Global configuration file.
.IP "~/.tudu.xml"
User default tudu.
.SH NO WARRANTIES
This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.
.SH SEE ALSO
.PP
TuDu Home Page: https://code.meskio.net/tudu/
.SH AUTHOR
TuDu was written by Rub\['e]n Poll\['a]n Bella <meskio@sindominio.net>.
.PP
This manual page was written by Rub\['e]n Poll\['a]n Bella and further
improved by Enrique Mat\['i]as S\['a]nchez.


ChangeLog
0.10.1
	* Fixed tmpnam warning
	* Fixed not null terminated strings

0.10
	* Added move task witch shift+up/down
	* Added mark task as done/undone with space
	* Fixed segfault on loading schedule

0.9.1
	* Added CPPFLAGS to the makefiles
	* Fixed a segfault adding sched
	* Fixed the cursor drawing on the editor

0.9
	* Added support for double with characters (like some chinese characters)
	* Added the default options on the man page
	* Fix a bug with some characters echoing on the editor

0.8.2
	* Added configuration variable 'tudu_file' to set the path to the xml file
	* Improved the category editor
	* Don't check configuration files when invoked with '-h' and '-v'
	* Fixed segfault on scroll help page when the resolution is big

0.8.1
	* Added arrow keys on date editor
	* Added support for non-latin keybinding pairs
	* Fixed wrong plotting while change tudu task position
	* Fixed category problems (autofill, showonly, ...)
	* Fixed Makefile problems reported on debian bug #611077
	* Removed freeze on moving task

0.8
	* Added ./configure script compatible with autotools
	* Added multiple category support
	* Added support for non-fix length categories
	* Added warning color to old scheduled tasks in sched
	* Added AvPag and RvPag to normal interface
	* Fixed display when the length of the title equal to the screen
	* Fixed problems with updating sched when the title changes

0.7
	* Added support to multi linear task titles
	* New pipe windows as separators
	* Configurable date format as US or european
	* Configurable tasks with child displayed bold
	* Warn when disk is full
	* Fixed install path problems if the DESTDIR didn't end with '/'
	* Fixed year xml storage with different locales (debian bug #568064)
	* Fixed problems with external editors and UTF-8
	* Fixed collision with some polish characters and control keys
	* Fixed dead loop when .tudu.xml is empty

0.6.1
	* Added confirm question on Q
	* Configurable beggining to end move
	* Fixed remaping keys on config file
	* Show remaped keys on help bar (Patch from Enrique Matias)
	* Fixed leap year calculus (Patch from Enrique Matias)
	* Make inline help and man more redeable (Patch from Enrique Matias)
	* Deactivate Ctrl-c (Patch from Thorbjörn Jemander)

0.6
	* Added suppor for international characters
	* Added help command to display man page
	* Added hide_percent option in the config file
	* Added patch from Magnus Müller: scroll to beginning end if end beginning reached
	* Fixed problems with O and P
	* Fixed problems at repaint after resize

0.5.1
	* Added a simple theme to show when theme don't fit
	* Added support for up/re page keys on text editor
	* Fixed tab problems on editors
	* Fixed problems with text cursor when resize window
	* Fixed problems with non writable files
	* Fixed default permisions on lock files
	* Fixed segfaults when wrong params in cli

0.5
	* Added search on titles
	* Added command interface
	* Added commands for hide categories
	* Added autocompletion on category editor
	* Added home and end keys on the editor
	* Added lock files for avoid concurrent tudus
	* Added documentation about all the features in the man page
	* Editing the title if is empty left arrow destroy the task
	* Fixed reorder tasks
	* Fixed english messages (thanks to Michael Leslie)

0.4.1
	* Added reorder tasks on the scheluder
	* Fixed paste as child when task is collapsed
	* Fixed refresh when sched is deleted

0.4
	* Added scheluder for tasks
	* Added hide done tasks
	* Added simple visual tree
	* Added paste as child
	* Sort by done uses 100% as done
	* Fixed help dirty close
	* Added sort after edit priority, category, ...

0.3.2
	* Fixed a segfault in i386 editing priorities
	* Added support for delete priority

0.3.1
	* Changed tudurc from /usr/local/share/tudu to /usr/local/etc

0.3
	* Added @include to config files
	* Added theme support in config files
	* Addeѕ support for resizing term
	* Added sort by percent done
	* Added scroll in help screen
	* Fixed sort errors
	* Fixed problems with temporal files
	* Fixed problems with empty tudus

0.2.1
	* Fixed bug: saving files in amd64
	* Fixed bug: inheritance deadlines
	* Fixed bug: normal text editor
	* Fixed bug: category editor

0.2
	* Added deadline date.
	* Added categories.
	* Added priorities.
	* Added support for sort the ToDos from different sort methods.
	* Allowed to edit the text of the ToDos with an external editor.
	* Fixed some segfaults and bugs.

0.1
	* First version.


src/tudu.cc

/**************************************************************************
 * Copyright (C) 2007-2015 Ruben Pollan Bella <meskio@sindominio.net>     *
 *                                                                        *
 *  This file is part of TuDu.                                            *
 *                                                                        *
 *  TuDu is free software; you can redistribute it and/or modify          *
 *  it under the terms of the GNU General Public License as published by  *
 *  the Free Software Foundation; version 3 of the License.        *
 *                                                                        *
 *  TuDu is distributed in the hope that it will be useful,               *
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *  GNU General Public License for more details.                          *
 *                                                                        *
 *  You should have received a copy of the GNU General Public License     *
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>. *
 **************************************************************************/


#include "includes.h"
extern int errno;
#include "interface.h"
#include "screen.h"
#include "data.h"
#include "sched.h"
#include "parser.h"
#include "config.h"

#define usage() \
	cout << "Usage: " << argv[0] << " [options]" << endl; \
	cout << "\t-f file\tload tudu file" << endl; \
	cout << "\t-c file\tload specific config file" << endl << endl; \
	cout << "\t-v\tshow version" << endl; \
	cout << "\t-h\tshow this usage message" << endl << endl; \
	cout << "The default config file is in ~/.tudurc" << endl;

#define version() \
	cout << VERSION_STR << endl; \
	cout << "TuDu Copyright (C) 2007-2015 Ruben Pollan Bella <meskio@sindominio.net>" << endl; \
	cout << "TuDu comes with ABSOLUTELY NO WARRANTY; for details type `tudu -vv'" << endl; \
	cout << "This is free software; you are welcome to redistribute it" << endl; \
	cout << "under certain conditions. Type `tudu -vv' for details." << endl;

#define copyright() \
	cout << VERSION_STR << endl; \
	cout << "Copyright (C) 2007-2015 Ruben Pollan Bella <meskio@sindominio.net>" << endl << endl; \
 	cout << "TuDu is free software; you can redistribute it and/or modify" << endl; \
 	cout << "it under the terms of the GNU General Public License as published by" << endl; \
 	cout << "the Free Software Foundation; version 3 of the License." << endl << endl; \
 	cout << "TuDu is distributed in the hope that it will be useful," << endl; \
 	cout << "but WITHOUT ANY WARRANTY; without even the implied warranty of" << endl; \
 	cout << "MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the" << endl; \
 	cout << "GNU General Public License for more details." << endl << endl; \
 	cout << "You should have received a copy of the GNU General Public License" << endl; \
	cout << "along with this program.  If not, see <http://www.gnu.org/licenses/>." << endl;

/* return true for delete lock file and false for close tudu */
bool lock_ask()
{
	cout << endl << "Lock file found" << endl;
	cout << endl << "either" << endl;
	cout << "    Another program may be editing the same file." << endl;
	cout << "    If this is the case, be careful not to end up with two" << endl;
	cout << "    different instances of the same file when making changes." << endl;
	cout << "    Quit, or continue with caution." << endl;
	cout << endl << "or" << endl;
	cout << "    A tudu session for this file crashed." << endl;
	cout << "    \"Edit anyway\" will destroy the lock file." << endl;
	cout << endl << "[E]dit anyway, [Q]uit:";
	string str;
       	cin >> str;

	if ((str[0] == 'E') || (str[0] == 'e'))
		return true;
	else
		return false;
}

int main(int argc, char **argv, char *env[])
{
	int i;
	char file_rc[128], file_xml[128], file_lock[133];

	// Disable Ctrl-C: Otherwise all changes are lost if one slips.
	struct sigaction ignore_ctrl_c;
	ignore_ctrl_c.sa_handler = SIG_IGN;
	ignore_ctrl_c.sa_flags = 0;
	sigaction(SIGINT, &ignore_ctrl_c, NULL);

	for (i = 0; strncmp(env[i],"HOME=",5); ++i);

	Config config;
	bool configErr = !config.load(CONFIG_FILE); // the error will be displayed after check args
	strncpy(file_rc,env[i]+5,119);
	file_rc[119] = '\0';
	strncat(file_rc,"/.tudurc", 9);
	config.load(file_rc);

	if (config.getTuduFile() == L"") {
		strncpy(file_xml,env[i]+5,117);
		file_xml[117] = '\0';
		strncat(file_xml,"/.tudu.xml", 10);
	} else {
		wcstombs(file_xml, config.getTuduFile().c_str(), 128);
	}

	/*
	 * Parse the comand line arguments
	 */
	for (i = 1; i < argc; ++i)
	{
		if (!strncmp("-f",argv[i],2))
		{
			++i;
			if (i < argc) {
				strncpy(file_xml, argv[i], 127);
				file_xml[127] = '\0';
			}
			else
			{
				usage();
				return 0;
			}
		}
		else if (!strncmp("-c",argv[i],2))
		{
			++i;
			if (i < argc)
			{
				if (argv[i][0] != '/')
				{
					int j;
					for (j = 0; strncmp(env[j],"PWD=",4); ++j);
					strncpy(file_rc,env[j]+4,99);
					file_rc[99] = '\0';
					strcat(file_rc,"/");
					strncat(file_rc,argv[i],27);
				}
				else
				{
					strncpy(file_rc, argv[i], 127);
					file_rc[127] = '\0';
				}
				config.load(file_rc);
			}
			else
			{
				usage();
				return 0;
			}
		}
		else if (!strncmp("-vv",argv[i],3))
		{
			copyright();
			return 0;
		}
		else if (!strncmp("-v",argv[i],2))
		{
			version();
			return 0;
		}
		else if (!strncmp("--help",argv[i],6) ||
				!strncmp("-h",argv[i],2))
		{
			usage();
			return 0;
		}
	}

	if (configErr)
	{
		fprintf(stderr, "Err: Global config does not exist. The config should be %s\n", CONFIG_FILE);
		exit(1);
	}

	/*
	 * Check and create the lock file
	 */
	strcpy(file_lock,file_xml);
	for (i = strlen(file_lock); (file_lock[i] != '/') && (i > 0); i--);
	if (file_lock[i] == '/') i++;
	if (file_lock[i] != '.')
	{
		file_lock[i] = '.';
		file_lock[i+1] = '\0';
		strcat(file_lock, file_xml+i);
	}
	strcat(file_lock,"_lock");
	// FIXME: it wont work with NFS
	int lock;
	lock = open(file_lock, O_CREAT|O_EXCL, 00666);
	if (lock == -1)
	{
		if (errno == EEXIST)
		{
			if (!lock_ask()) exit(1);
		}
		/* no rights to write in this folder skip the problem */
		else if (errno != EACCES)
		{
			fprintf(stderr, "Err: I can not create the lock file %s\n", file_lock);
			exit(1);
		}
	}
	else
	{
		close(lock);
	}

	/*
	 * Load data
	 */
	ToDo node;
	iToDo it(node);
	Sched sched;
	Parser p(file_xml);
	if (!p.parse(node,sched))
	{
		Parser welcome(WELCOME_FILE);
		/* welcome file don't exist */
		if (!welcome.parse(node,sched))
		{
			fprintf(stderr, "Err: Welcome file does not exist. It should be %s\n", WELCOME_FILE);
			unlink(file_lock);
			exit(1);
		}
	}

	/*
	 * Load end start interface
	 */
	Writer w(file_xml,node);
	Screen screen(config);
	Cmd cmd;
	Interface in(screen,it,sched,config,w,cmd);
	in.main();

	/*
	 * Delete lock file
	 */
	unlink(file_lock);
	return 0;
}
